File: back-end/src/main/java/nl/hva/backend/domain/models/game/board/GameBoard.java
package nl.hva.backend.domain.models.game.board;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.board.location.Location;
import nl.hva.backend.domain.models.game.card.Card;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Entity

@Getter
@Setter
@NoArgsConstructor

@JsonIdentityInfo(
        generator = ObjectIdGenerators.PropertyGenerator.class,
        property = "id",
        scope = GameBoard.class
)
public class GameBoard {
    public final static int FIRST_POSITION = 1;
    public final static int LAST_POSITION = 40;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    private String name;

    @OneToMany(mappedBy = "gameBoard", fetch = FetchType.EAGER)
    private List<Location> locations;

    @OneToMany(mappedBy = "gameBoard")
    private List<Card> cards;

    public GameBoard(String name) {
        this.name = name;
    }

    public void add(Location location) {
        if (location == null) return;

        if (getLocations() == null || !(getLocations() instanceof ArrayList))
            setLocations(new ArrayList<>(Objects.requireNonNullElse(getLocations(), List.of())));

        getLocations().add(location);
    }

    public void add(Card card) {
        if (card == null) return;

        if (getCards() == null || !(getCards() instanceof ArrayList))
            setCards(new ArrayList<>(Objects.requireNonNullElse(getCards(), List.of())));

        getCards().add(card);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof GameBoard gameBoard)) return false;
        return getId() == gameBoard.getId();
    }

    @Override
    public int hashCode() {
        return Long.hashCode(getId());
    }

}

File: front-end/src/app/domain/models/game/board/GameBoard.ts
import {Location} from "./location/Location";
import {LocationUtil} from "./location/LocationUtil";
import {Card} from "../card/Card";
import {CardUtil} from "../card/CardUtil";

export class GameBoard {
  public static readonly FIRST_POSITION: number = 1;
  public static readonly LAST_POSITION: number = 40;

  private _id: number;
  private _name: string;
  private _locations: Location[];

  public constructor(
    id: number,
    name: string,
    locations: Location[],
    cards: Card[]
  ) {
    this._id = id;
    this._name = name;
    this._locations = locations;
    this._cards = cards;
  }

  private _cards: Card[];

  public get cards(): Card[] {
    return this._cards;
  }

  public get id(): number {
    return this._id;
  }

  public get name(): string {
    return this._name;
  }

  public get locations(): Location[] {
    return this._locations;
  }

  public static trueCopy(gameboard: GameBoard): GameBoard {
    return new GameBoard(
      gameboard.id,
      gameboard.name,
      gameboard.locations.map(location => LocationUtil.trueCopy(location)),
      gameboard.cards.map(card => CardUtil.trueCopy(card))
    );
  }

  public toJson(): object {
    return {
      "id": this.id,
      "name": this.name,
      "locations": this.locations.map(location => location.toJson()),
      "cards": this.cards.map(card => card.toJson())
    }
  }

}
